lib/rex/socket.rb

```diff
@@ -809,10 +828,64 @@ def type?
 
 protected
 
-  attr_writer :peerhost, :peerport, :localhost, :localport # :nodoc:
+  attr_writer :peerhost, :peerhostname, :peerport, :localhost, :localport # :nodoc:
   attr_writer :context # :nodoc:
   attr_writer :ipv # :nodoc:
 
+  def self.rex_gethostbyname(name, resolver = @@resolver)
+    raise ::SocketError.new(
+      "Rex::Socket internal DNS resolution requires passing/setting a resolver"
+    ) unless resolver
+    # Pull both record types
+    v4 = resolver.send(name, ::Net::DNS::A).answer.first
+    v6 = resolver.send(name, ::Net::DNS::AAAA).answer.first
+    # Emulate ::Socket's error if no responses found
+    if !v4 and !v6
+      raise ::SocketError.new('getaddrinfo: Name or service not known')
+    end
+    # Build response array
+    hostbyname = [name, []]
+    if v4
+      hostbyname << ::Socket::AF_INET
+      hostbyname << self.addr_aton(v4.address.to_s)
+      hostbyname << self.addr_aton(v6.address.to_s) if v6
```

Comment:
This assumes that the `address` fields are correctly populated from earlier. Might be good to check their contents prior to casting like this.

Reply:
Do we want to type-check with `is_a?` or the like?

Reply:
Yep pretty much. Just wanted to check that it wasn't blank and we are trying to act on a blank response. So something like `.blank?` and a check for the type here would be good?

Reply:
This is negated in part by this code:

```
    # Emulate ::Socket's error if no responses found
    if !v4[0] and !v6[0]
      raise ::SocketError.new('getaddrinfo: Name or service not known')
    end
```

This is in `  def self.rex_resolve_hostname(name, resolver = @@resolver)` inside `lib/rex/socket.rb`.